tests/test_classical_gradients.py

```diff
@@ -15,129 +15,196 @@
-
-    def test_gradient_multiargs(self):
-        """Tests gradients of univariate functions with multiple arguments in signature."""
-        self.logTestName()
-
-        for gradf, f, name in zip(self.grad_margs_funcs, self.margs_fns, self.fnames):
-            with self.subTest(i=name):
-                for jdx in range(len(x_vals[:-1])):
-                    x = x_vals[jdx]
-                    y = x_vals[jdx + 1]
-                    # gradient wrt first argument
-                    gx = qml.grad(f, 0)
-                    auto_gradx = gx(x,y)
-                    correct_gradx = gradf(x,y)[0]
-                    self.assertAllAlmostEqual(auto_gradx, correct_gradx, delta=self.tol)
-                    # gradient wrt second argument
-                    gy = qml.grad(f, 1)
-                    auto_grady = gy(x,y)
-                    correct_grady = gradf(x,y)[1]
-                    self.assertAllAlmostEqual(auto_grady, correct_grady, delta=self.tol)
-                    # gradient wrt both arguments
-                    gxy = qml.grad(f, [0,1])
-                    auto_gradxy = gxy(x,y)
-                    correct_gradxy = gradf(x,y)
-                    self.assertAllAlmostEqual(auto_gradxy, correct_gradxy, delta=self.tol)
-
-    def test_gradient_multivar(self):
-        """Tests gradients of multivariate unidimensional functions."""
-        self.logTestName()
-
-        for gradf, f, name in zip(self.grad_multi_funcs, self.multivariate_funcs, self.fnames):
-            with self.subTest(i=name):
-                for jdx in range(len(x_vals[:-1])):
-                    x_vec = x_vals[jdx:jdx+2]
-                    g = qml.grad(f, 0)
-                    auto_grad = g(x_vec)
-                    correct_grad = gradf(x_vec)
-                    self.assertAllAlmostEqual(auto_grad, correct_grad, delta=self.tol)
-
-    def test_gradient_multivar_multidim(self):
-        """Tests gradients of multivariate multidimensional functions."""
-        self.logTestName()
-
-        for gradf, f, name in zip(self.grad_mvar_mdim_funcs, self.mvar_mdim_funcs, self.fnames):
-            with self.subTest(i=name):
-                for jdx in range(len(x_vals[:-1])):
-                    x_vec = x_vals[jdx:jdx+2]
-                    x_vec_multidim = np.expand_dims(x_vec, axis=1)
-                    g = qml.grad(f, 0)
-                    auto_grad = g(x_vec_multidim)
-                    correct_grad = gradf(x_vec_multidim)
-                    self.assertAllAlmostEqual(auto_grad, correct_grad, delta=self.tol)
-
-
-if __name__ == '__main__':
-    print('Testing PennyLane version ' + qml.version() + ', classical gradients.')
-    # run the tests in this file
-    suite = unittest.TestSuite()
-    for t in (BasicTest,):
-        ttt = unittest.TestLoader().loadTestsFromTestCase(t)
-        suite.addTests(ttt)
-
-    unittest.TextTestRunner().run(suite)
+
+def test_gradient_univar(tol):
```

Comment:
To address your comment on multiple tests per test function, what you could do here is make each of the existing test functions a _class_, and then break the contents of each class into test methods, e.g.,

```python
class TestGradientUnivar:
    """Tests gradients of univariate unidimensional functions calculated via qml.grad."""

    def test_sin(self, tol):
        """tests that the gradient of the sin function is calculated correctly"""
        x_vals = np.linspace(-10, 10, 16, endpoint=False)
        g = qml.grad(np.sin, 0)
        auto_grad = [g(x) for x in x_vals]
        correct_grad = np.cos(x_vals)
        assert np.allclose(auto_grad, correct_grad, atol=tol, rtol=0)
```

Reply:
Thanks for the suggestion. The tests look much better now. This is ready for review.